### ISSUE TITLE
pinvoke of native vararg functions causes "invalid IL" error

### ISSUE THREAD

**Issue by zfigura** [mono/mono#15411]
Hello folks,

I encountered this problem while debugging an application through wine-mono.

The problem is essentially that attempting to pinvoke a native vararg function uses the wrong signature: it uses the signature from its MethodDef parent instead of the signature of the function itself. As a result, it doesn't pick up any of the actual vararg arguments, but only the always-present arguments which are given in the MethodDef table. Thus mono_method_to_ir() gets the wrong number of arguments, which inevitably causes problems as there are more objects in the stack than there should be.

I'm certainly willing to try to fix this myself, but I'm unsure of some things. This is largely because I'm quite unfamiliar with Mono's code base, but also because I'm a bit unsure which of two potential approaches to pursue:
1. Construct the "real" method in method_from_memberref (or thereabouts), i.e. copy the flags from the MethodDef parent, but set the token to the real method's token. This seems like the most correct thing to do to me, because we key not just the signature on the MonoMethod, but also the inlining info (which I think is going to depend on the signature, unless I'm misreading the code), and anything else that relies on the token, including the wrapper. However, this is also tricky, because a lot of things rely on this function, and multiple places rely specifically on getting the MethodDef parent, e.g. mono_method_get_header_internal().
2. Deal with the difference at a higher level, i.e. provide functions for retrieving the method signature and wrapper (and presumably other things) from the correct token, and use them in mono_method_to_ir(). This seems more complex to me, but it's what we already do in mono_method_get_signature_checked().

Both of these will likely require a lot of research, so I'd appreciate some help first.

## Steps to Reproduce

1. Apply the attached file against Mono.
2. Compile the testsuite.
3. Run vararg3.exe.

[mono_native_vararg_test.diff.txt](https://github.com/mono/mono/files/3322858/mono_native_vararg_test.diff.txt)

### Current Behavior

The test crashes, with:

    Unhandled Exception:
    System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.InvalidProgramException: Invalid IL code in Tests:test_0_varargs_unmanaged (): IL_0008: ret

### Expected Behavior

The test runs successfully, as with Microsoft's .NET.

## On which platforms did you notice this

[ ] macOS
[x] Linux
[ ] Windows

**Version Used**: mono-3.2.3-25296-g5ca92bdbe8f

**Comment by vargaz** [comment]
Yes, this is not supported right now. You might be able to work around it by defining a separate c# signature for each possible vararg signature while keeping the dllimport entrypoint the same.

**Comment by zfigura** [comment]
> Yes, this is not supported right now. You might be able to work around it by defining a separate c# signature for each possible vararg signature while keeping the dllimport entrypoint the same.

Unfortunately that won't work for my use case, I'm trying to run an already compiled IL through wine-mono.

**Comment by zfigura** [comment]
After becoming more familiar with Mono source, it does seem more like #2 is the correct way to fix this. I'm guessing that we want to pass an extra parameter to mono_marshal_get_native_wrapper() specifying the signature. Is that correct? Do we want to do something else entirely?

That said, I'm not sure how to deal with functions that don't know the correct signature, e.g. compile_special() in mini/mini-runtime.c. Since we trampoline to that function (ultimately) when executing a 'call' instruction, and it calls the pinvoke's wrapper (I think? this part of the code is not particularly easy to gain a high-level understanding of) it needs to be able to retrieve a/the wrapper with the right signature.

**Comment by madewokherd** [comment]
It seems like getting the correct signature in mono_method_to_ir is the easy part, we can just use mono_method_get_signature_checked as is already done for normal vararg calls. The hard part is making sure we get a correct wrapper for the MemberRef.

It seems like the trampoline code isn't designed to pass a signature or token around, which makes me suspect that a method for the MemberRef is the way to go. We'd want to store this method in a table for the image and create it when we encounter a call to a pinvoke vararg function with optional arguments, but we wouldn't want to run it through mono_marshal_get_native_wrapper just yet. So what we'd actually want is a version of the pinvoke method as it would exist in the image, if an image were allowed to have methods with a sentinel in the signature, which would get eventually passed to mono_marshal_get_native_wrapper in the usual way.

We need to preserve the sentinel because it impacts the calling convention. At the very least, anything after the sentinel needs to be on the stack. This is going to require architecture-specific changes.

The interpreter will need handling for this as well, and I don't know how that has to work.

**Comment by madewokherd** [comment]
Apparently I'm wrong about the sentinel impacting the calling convention, at least on amd64. It seems the optional arguments go in registers as usual.

Can we assume that's true of all architectures? It would simplify things a little.

**Comment by madewokherd** [comment]
Searching gave me some GCC documentation that seems to think it may matter for some architectures. https://gcc.gnu.org/onlinedocs/gnat_ugn/Calling-Conventions.html
